File: src/main/java/kata/MeetupSubscribe.java
package kata;

import kata.persistence.MeetupEventRepository;

import java.time.Instant;
import java.time.LocalDateTime;
import java.util.List;

import static java.util.stream.Collectors.toList;

public class MeetupSubscribe {

    private final MeetupEventRepository repository;

    public MeetupSubscribe(MeetupEventRepository repository) {
        this.repository = repository;
    }

    public Long registerMeetupEvent(String eventName, Integer eventCapacity, LocalDateTime startTime) {
        var id = repository.generateId();
        var meetupEvent = new MeetupEvent(id, eventCapacity, eventName, startTime, List.of());

        repository.save(meetupEvent);

        return id;
    }

    public void subscribeUserToMeetupEvent(String userId, Long meetupEventId) {
        var meetup = repository.findById(meetupEventId);

        var userSubscription = meetup.getSubscription(userId);
        if (userSubscription != null) {
            throw new RuntimeException(String.format("User %s already has a subscription", userId));
        }

        var addToWaitingList = meetup.getParticipants().size() == meetup.getCapacity();
        var subscription = new Subscription(userId, Instant.now(), addToWaitingList);
        meetup.getSubscriptions().add(subscription);

        repository.save(meetup);
    }

    public void cancelUserSubscription(String userId, Long meetupEventId) {
        var meetup = repository.findById(meetupEventId);

        var inWaitingList = meetup.isInWaitingList(userId);
        meetup.remove(userId);

        if (!inWaitingList) {
            List<Subscription> waitingList = meetup.getWaitingList();
            if (!waitingList.isEmpty()) {
                Subscription firstInWaitingList = waitingList.get(0);
                meetup.changeFromWaitingListToParticipant(firstInWaitingList.getUserId());
            }
        }

        repository.save(meetup);
    }

    public void increaseCapacity(Long meetupEventId, int newCapacity) {
        var meetupEvent = repository.findById(meetupEventId);

        var oldCapacity = meetupEvent.getCapacity();
        if (oldCapacity < newCapacity) {
            var updatedMeetupEvent = meetupEvent.withCapacity(newCapacity);
            var newSlots = newCapacity - oldCapacity;
            meetupEvent.getWaitingList()
                    .stream()
                    .limit(newSlots)
                    .forEach(subscription -> meetupEvent.changeFromWaitingListToParticipant(subscription.getUserId()));

            repository.save(updatedMeetupEvent);
        }
    }

    public MeetupEventStatusDto getMeetupEventStatus(Long meetupEventId) {
        var meetupEvent = repository.findById(meetupEventId);

        var meetupEventStatusDto = new MeetupEventStatusDto();
        meetupEventStatusDto.meetupId = meetupEvent.getId();
        meetupEventStatusDto.eventCapacity = meetupEvent.getCapacity();
        meetupEventStatusDto.startTime = meetupEvent.getStartTime();
        meetupEventStatusDto.eventName = meetupEvent.getEventName();
        meetupEventStatusDto.waitingList = meetupEvent.getWaitingList().stream().map(Subscription::getUserId).collect(toList());
        meetupEventStatusDto.participants = meetupEvent.getParticipants().stream().map(Subscription::getUserId).collect(toList());
        return meetupEventStatusDto;
    }
}

File: src/main/kotlin/kata/MeetupSubscribe.kt
package kata

import kata.persistence.MeetupEventRepository
import java.time.LocalDateTime
import java.lang.RuntimeException
import java.time.Clock
import java.time.Instant
import java.util.stream.Collectors

class MeetupSubscribe(
  private val repository: MeetupEventRepository,
  private val clock: Clock,
) {

  fun registerMeetupEvent(eventName: String, eventCapacity: Int, startTime: LocalDateTime): Long {
    val id = repository.generateId()
    val meetup = newMeetup(id, eventName, eventCapacity, startTime)
    repository.save(meetup)
    return id
  }

  fun subscribeUserToMeetupEvent(userId: String, meetupEventId: Long) {
    val meetup = repository.findById(meetupEventId)

    if (meetup.state.hasSubscriptionFor(userId)) {
      throw RuntimeException(String.format("User %s already has a subscription", userId))
    }

    val updateMeetup = meetup.subscribe(userId, Instant.now(clock))
    repository.save(updateMeetup)
  }

  fun cancelUserSubscription(userId: String, meetupEventId: Long) {
    val meetup = repository.findById(meetupEventId)

    val updatedMeetup = meetup.unsubscribe(userId, meetupEventId)

    repository.save(updatedMeetup)
  }

  fun increaseCapacity(meetupEventId: Long, newCapacity: Int) {
    val meetupEvent = repository.findById(meetupEventId)

    val oldCapacity = meetupEvent.state.capacity
    if (oldCapacity < newCapacity) {
      val updatedMeetup = meetupEvent.increaseCapacityTo(newCapacity)
      repository.save(updatedMeetup)
    }
  }

  fun getMeetupEventStatus(meetupEventId: Long): MeetupEventStatusDto {
    val meetupEvent = repository.findById(meetupEventId).state

    return MeetupEventStatusDto(
      meetupId = meetupEvent.id,
      eventCapacity = meetupEvent.capacity,
      startTime = meetupEvent.startTime,
      eventName = meetupEvent.eventName,
      waitingList = meetupEvent.waitingList.stream().map { obj: Subscription -> obj.userId }.collect(Collectors.toList()),
      participants = meetupEvent.participants.stream().map { obj: Subscription -> obj.userId }.collect(Collectors.toList()),
    )
  }
}
